transcom/mymove
# B-20674 TPPS Paid Info INT

> [!IMPORTANT]  
> [Upstream](https://github.com/transcom/mymove/pull/13434).

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A982872)

## Summary

When there is data in the `tpps_paid_invoice_reports` table as a result of processing a TPPS Paid Invoice Report from US bank, we want to display that on the payment request for the TIO to view.

- pulled in B-20671 work
- added a foreign key to the `tpps_paid_invoice_reports` table so that the `invoice_number` from the TPPS report will reference the `payment_requests.payment_request_number` (which has a unique constraint)
- added in the fetching of relevant TPPS data for the payment request
- added the total amount TPPS paid and the seller paid date to the payment request payload
- added the total amount TPPS paid per service item to the payment service item payload
- displayed those values on the frontend and added client tests

### How to test

1. Access the office app as a TIO
2. Pick a payment request needing reviewed that has DLH (domestic linehaul), FSC (fuel surcharge), DDP (domestic destination price), and DUPK (domestic unpacking service items. I used DSTSIT.
![image](https://github.com/user-attachments/assets/5bf65e23-e6ee-4a38-8a26-8350e8854b07)

3. Review the payment request, rejecting a few service items like the move management and counseling ones, while accepting the rest. You want to make sure you have both an accepted and a rejected amount.
![image](https://github.com/user-attachments/assets/9f525b88-f41a-4abc-b867-0b86b1553023)

4. For the payment request you've been working with, grab the payment request number.
![image](https://github.com/user-attachments/assets/c86336ef-422b-4041-bb8d-f709e82bfd62)

5. You're going to insert some fake data into the `tpps_paid_invoice_reports` table and you want the `invoice_number` for those entries to match your payment request number.
6. **Modify the following script so that the `invoice_number` values for all 4 entries match your payment request number**. You might also have to mess with the `id` uuids.
```
insert into tpps_paid_invoice_reports 
-- 1
VALUES ('b064d62a-9576-402a-b806-41933339c746', -- id
'9567-5270-1', --invoice_number
'2024-07-29 00:00:00.000', -- tpps_created_doc_date
'2024-07-30 00:00:00.000', -- seller_paid_date
115155000, -- invoice_total_charges_in_millicents
'DDP', -- line_description
'DDP', -- product_description
'3760', -- line_billing_units
'770', --line_unit_price_in_millicents
'2895000', -- line_net_charge_in_millicents
'1841-7267-826285fc', -- po_tcn
'1', -- line_number
'INT', -- first_note_code
'Notes to My Company - INT', -- first_note_description
'CARR', -- first_note_to
'HQ50066', -- first_note_message
'', -- second_note_code
'', -- second_note_description
'', -- second_note_to
'', -- second_note_message
'', -- third_note_code
'', -- third_note_code_description
'', -- third_note_code_to
'', -- third_note_code_message
'2024-08-01 00:00:00.000', -- created_at
'2024-08-01 00:00:00.000'), -- updated_at
-- 2
('b064d62a-9576-402a-b806-41933339c747', -- id
'9567-5270-1', --invoice_number
'2024-07-29 00:00:00.000', -- tpps_created_doc_date
'2024-07-30 00:00:00.000', -- seller_paid_date
115155000, -- invoice_total_charges_in_millicents
'FSC', -- line_description
'FSC', -- product_description
'3760', -- line_billing_units
'140', --line_unit_price_in_millicents
'539000', -- line_net_charge_in_millicents
'1841-7267-aeb3cfea', -- po_tcn
'4', -- line_number
'INT', -- first_note_code
'Notes to My Company - INT', -- first_note_description
'CARR', -- first_note_to
'HQ50066', -- first_note_message
'', -- second_note_code
'', -- second_note_description
'', -- second_note_to
'', -- second_note_message
'', -- third_note_code
'', -- third_note_code_description
'', -- third_note_code_to
'', -- third_note_code_message
'2024-08-01 00:00:00.000', -- created_at
'2024-08-01 00:00:00.000'), -- updated_at
-- 3
('b064d62a-9576-402a-b806-41933339c748', -- id
'9567-5270-1', --invoice_number
'2024-07-29 00:00:00.000', -- tpps_created_doc_date
'2024-07-30 00:00:00.000', -- seller_paid_date
115155000, -- invoice_total_charges_in_millicents
'DLH', -- line_description
'DLH', -- product_description
'3760', -- line_billing_units
'26560', --line_unit_price_in_millicents
'99877000', -- line_net_charge_in_millicents
'1841-7267-c8ea170b', -- po_tcn
'2', -- line_number
'INT', -- first_note_code
'Notes to My Company - INT', -- first_note_description
'CARR', -- first_note_to
'HQ50066', -- first_note_message
'', -- second_note_code
'', -- second_note_description
'', -- second_note_to
'', -- second_note_message
'', -- third_note_code
'', -- third_note_code_description
'', -- third_note_code_to
'', -- third_note_code_message
'2024-08-01 00:00:00.000', -- created_at
'2024-08-01 00:00:00.000'), -- updated_at
-- 4
('b064d62a-9576-402a-b806-41933339c749', -- id
'9567-5270-1', --invoice_number
'2024-07-29 00:00:00.000', -- tpps_created_doc_date
'2024-07-30 00:00:00.000', -- seller_paid_date
115155000, -- invoice_total_charges_in_millicents
'DUPK', -- line_description
'DUPK', -- product_description
'3760', -- line_billing_units
'3150', --line_unit_price_in_millicents
'11844000', -- line_net_charge_in_millicents
'1841-7267-265c16d7', -- po_tcn
'3', -- line_number
'INT', -- first_note_code
'Notes to My Company - INT', -- first_note_description
'CARR', -- first_note_to
'HQ50066', -- first_note_message
'', -- second_note_code
'', -- second_note_description
'', -- second_note_to
'', -- second_note_message
'', -- third_note_code
'', -- third_note_code_description
'', -- third_note_code_to
'', -- third_note_code_message
'2024-08-01 00:00:00.000', -- created_at
'2024-08-01 00:00:00.000'
);
```
7. Make sure the `tpps_paid_invoice_reports` table has the data you just inserted
![image](https://github.com/user-attachments/assets/90c7e1b6-287b-4242-b5b1-0f8a98b9e098)

8. Update the payment request you've been using from a status of REVIEWED to a status of PAID
![image](https://github.com/user-attachments/assets/5817a4c7-14b9-4334-8b16-5140f59050ac)

9. Refresh your Payment Requests page.
10. Verify that you see the `TPPS PAID` status flag next to the payment request number
![image](https://github.com/user-attachments/assets/1da81b1d-c1ef-4e00-83c7-85c623dba79a)

11. Verify that you see the accepted, rejected, and TPPS paid total values at the top, under the payment request number
![image](https://github.com/user-attachments/assets/6375467d-e2a4-4dfc-b947-eca1c8ec8743)

12. Expand the show request details drop down.
13. Verify that you see a new column of TPPS paid with values for the entries that you inserted into the database
```
DLH - $998.77
FSC - $5.39
DDP - $28.95
DUPK - $118.44
```
![image](https://github.com/user-attachments/assets/11a0e6c1-3e31-40e8-b33e-1fdb028a230a)

14. Go back to the TIO queue and pick a new payment request. Make sure you don't see the TPPS paid column on the details for the service items - you should only see that column and values there when there is TPPS data to be shown.

![image](https://github.com/user-attachments/assets/f41a93a8-1c9f-42d4-845e-23cf0e9d910b)
![image](https://github.com/user-attachments/assets/391e04a5-130f-4ab6-911e-6a8f463f63fa)
